### PR TITLE
[SPARK-45617][BUILD] Upgrade Apache Commons Crypto 1.2.0

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch upgrades Apache Commons Crypto to 1.2.0.

### Why are the changes needed?

Currently used 1.1.0 is more than 3 years ago (2020-08-28 [released](https://commons.apache.org/proper/commons-crypto/changes-report.html)).

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: dev/deps/spark-deps-hadoop-3-hive-2.3 [hunk 1]
```diff
@@ -118,6 +119,7 @@ java-diff-utils/4.12//java-diff-utils-4.12.jar
 javassist/3.29.2-GA//javassist-3.29.2-GA.jar
 javax.jdo/3.2.0-m3//javax.jdo-3.2.0-m3.jar
 javolution/5.5.1//javolution-5.5.1.jar
+jaxb-impl/2.3.7//jaxb-impl-2.3.7.jar
```
Comment (dongjoon-hyun) [inline dev/deps/spark-deps-hadoop-3-hive-2.3:122]: Oh, may I ask why this is introduced at 1.2.0 newly?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. I have only one question about the newly added dependency.

File: dev/deps/spark-deps-hadoop-3-hive-2.3 (hunk 1 above)
Comment (viirya) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:122]: I'm not sure if it is brought by Apache Commons Crypto 1.2.0 or it is just not updated. You know something this file is forgot to update.

Comment (viirya) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:122]: Let me check it.

Comment (viirya) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:122]: Ok, it is brought by this upgrade.

Comment (viirya) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:122]: In the release [notes](https://commons.apache.org/proper/commons-crypto/changes-report.html), it shows the release bumps jaxb-impl from 2.3.6 to 2.3.7. Not sure why jaxb-impl 2.3.6 is not shown before for 1.1.0.

**Review by LuciferYang — CHANGES_REQUESTED** [review]
@viirya @dongjoon-hyun Maybe we can't upgrade this dependency. we upgraded it in SPARK-42488, but revert it back later due to test failures on Apple Silicon. Please see the details at https://github.com/apache/spark/pull/40082#issuecomment-1443129166

**Comment by LuciferYang** [comment]
I have tried again, using Crypto 1.2.0 in `RowQueueSuite` on Apple Silicon still results in a test failure:

```


[info] RowQueueSuite:
[info] - in-memory queue (25 milliseconds)
[info] - disk queue (encryption = off) (61 milliseconds)
[info] org.apache.spark.sql.execution.python.RowQueueSuite *** ABORTED *** (130 milliseconds)
[info]   java.lang.ExceptionInInitializerError:
[info]   at java.base/java.lang.Class.forName0(Native Method)
[info]   at java.base/java.lang.Class.forName(Class.java:467)
[info]   at org.apache.commons.crypto.utils.ReflectionUtils.getClassByNameOrNull(ReflectionUtils.java:93)
[info]   at org.apache.commons.crypto.utils.ReflectionUtils.getClassByName(ReflectionUtils.java:64)
[info]   at org.apache.commons.crypto.random.CryptoRandomFactory.getCryptoRandom(CryptoRandomFactory.java:189)
[info]   at org.apache.spark.security.CryptoStreamUtils$.createInitializationVector(CryptoStreamUtils.scala:138)
[info]   at org.apache.spark.security.CryptoStreamUtils$.createCryptoOutputStream(CryptoStreamUtils.scala:56)
[info]   at org.apache.spark.serializer.SerializerManager.$anonfun$wrapForEncryption$3(SerializerManager.scala:151)
[info]   at scala.Option.map(Option.scala:242)
[info]   at org.apache.spark.serializer.SerializerManager.wrapForEncryption(SerializerManager.scala:151)
[info]   at org.apache.spark.sql.execution.python.DiskRowQueue.<init>(RowQueue.scala:119)
[info]   at org.apache.spark.sql.execution.python.RowQueueSuite.$anonfun$new$2(RowQueueSuite.scala:72)
[info]   at org.apache.spark.sql.execution.python.RowQueueSuite.$anonfun$new$2$adapted(RowQueueSuite.scala:68)
[info]   at org.apache.spark.security.EncryptionFunSuite.$anonfun$encryptionTest$2(EncryptionFunSuite.scala:32)
...
[info] Run completed in 1 second, 335 milliseconds.
[info] Total number of tests run: 2
[info] Suites: completed 0, aborted 1
[info] Tests: succeeded 2, failed 0, canceled 0, ignored 0, pending 0
[info] *** 1 SUITE ABORTED ***
[error] Error during tests:
[error]         org.apache.spark.sql.execution.python.RowQueueSuite
[error] (sql / Test / testOnly) sbt.TestsFailedException: Tests unsuccessful

```

**Comment by viirya** [comment]
Oh, I see. `commons-crypto-1.2.0.jar` only contain x86_64 Mac native library, it doesn't provide pre-built native library for Apple Silicon.

Hmm, but I also checked `commons-crypto-1.1.0.jar` which Spark uses currently. It also doesn't provide native library for Apple Silicon..

But `RowQueueSuite` passes with `commons-crypto-1.1.0` on M1 Mac.

**Comment by viirya** [comment]
It's weird. I will go to open a JIRA ticket [CRYPTO-169](https://issues.apache.org/jira/browse/CRYPTO-169). Close this for now.

**Comment by viirya** [comment]
Thanks @dongjoon-hyun @LuciferYang

**Comment by viirya** [comment]
Interesting. According to the replies on [CRYPTO-169](https://issues.apache.org/jira/browse/CRYPTO-169), Commons Crypto has never supported Apple silicon.

**Comment by LuciferYang** [comment]
Yes, it has never been supported. However, this does not prevent Apple Silicon from using `JavaCryptoRandom`. The original logic was that as long as either `OpenSslCryptoRandom` or `JavaCryptoRandom` initializes successfully, it would be acceptable. But in 1.2.0, if `OpenSslCryptoRandom` fails to initialize, it directly fails.

I suspect the change in behavior comes from this logic added in 1.2.0: https://github.com/apache/commons-crypto/blob/c5d19d017d96e34a4589736aeb48bc1d59e204a6/src/main/java/org/apache/commons/crypto/random/OpenSslCryptoRandom.java#L62-L72

image

The `initException` mechanism should be reused here instead of throwing an `RuntimeException`, because `RuntimeException` will also be wrapped into `ExceptionInInitializerError` by Java's own mechanism. This will cause the upper layer in 1.20 to be unable to handle this Error and continue to throw it up, which is the failure we are seeing now. 

image


However, since the `commons-crypto` project cannot be compiled on Apple Silicon, and I currently only have Apple Silicon development machines at hand, I am temporarily unable to verify this guess.

**Comment by LuciferYang** [comment]
> Interesting. According to the replies on [CRYPTO-169](https://issues.apache.org/jira/browse/CRYPTO-169), Commons Crypto has never supported Apple silicon.

Fixed: https://github.com/apache/commons-crypto/pull/258

**Comment by viirya** [comment]
Thanks. We can wait for 1.2.1 release then.

**Comment by LuciferYang** [comment]
I discovered a reentrancy issue during the locally testing, I am trying to further fix it ...

https://github.com/apache/commons-crypto/pull/259

**Pull request closed without merging**